### PR TITLE
Data Wrangling: Twitter, DaNews, DAGW

### PR DESCRIPTION
**Pull request by KennethEnevoldsen**

### CODE REVIEW

**Comment by MartinBernstorff** [comment]
@KennethEnevoldsen Anything in particular you would like me to take a look at? :-)

**Review by MartinBernstorff — COMMENTED** [review]
Minor readability feedback, and one possible bug. 

I'm not confident in my reading of the twitter parsing, so I trust you've done enough tests on real data to be confident in it :-)

Lots of good work, nicely done!

File: src/dfm/projects/data_processing/convert_dagw_to_jsonlgz.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        "naat": "Conversation",
+        "opensub": "Conversation",
+        "relig": "Wiki & Books",
+        "spont": "Conversation",
+        "synne": "Other",
+        "tv2r": "News",
+        "wiki": "Wiki & Books",
+        "wikibooks": "Wiki & Books",
+        "wikisource": "Wiki & Books",
+        "twfv19": "Social Media",  # not present in this version of the dataset
```
[the remark this answers is not in the document]
Comment (KennethEnevoldsen) [reply src/dfm/projects/data_processing/convert_dagw_to_jsonlgz.py:49]: Nope. Just wanted it to be human readable

File: src/dfm/projects/data_processing/convert_dagw_to_jsonlgz.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    "created": "..."         # OPTIONAL: timestamp when orig document was created (best-guess if not available)
+    "metadata": {...}        # OPTIONAL: source-specific metadata
+}
+"""
+
+from datasets import Dataset, DatasetDict, load_dataset  # type: ignore
+
+
+def reformat_dataset(ds: Dataset, num_proc: int) -> Dataset:
+    # current keys: dict_keys(['text', 'source', 'doc_id', 'LICENSE', 'uri', 'date_built'])
```
[the remark this answers is not in the document]
Comment (KennethEnevoldsen) [reply src/dfm/projects/data_processing/convert_dagw_to_jsonlgz.py:18]: Yep det er bare eksisterende kolonner

File: src/dfm/projects/data_processing/convert_dagw_to_jsonlgz.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        "adl": " 1700-2021",
+        "botxt": "contemporary (2021)",
+        "danavis": "1999-2003",
+        "dannet": "contemporary (2021)",
+        "depbank": "contemporary (2021)",
+        "ep": "2004-2008",
+        "ft": "2009-2019",
+        "gutenberg": "1700-now",
+        "jvj": "-",
+        "naat": "1930-now",
```
[the remark this answers is not in the document]
Comment (KennethEnevoldsen) [reply src/dfm/projects/data_processing/convert_dagw_to_jsonlgz.py:73]: Yes we discussed it. I Should correct this to match.

File: src/dfm/projects/data_processing/convert_danews_to_jsonlgz.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    """
+    text: str = ""
+    text += f"# {example['Heading']}\n\n"
+    if example["SubHeading"]:
+        text += f"## {example['SubHeading']}\n\n"
+
+    if example["Lead"]:
+        text += example["Lead"] + "\n"  # type: ignore
+    text += example["BodyText"] + "\n"  # type: ignore
+    text = text.strip()  # type: ignore
```
[the remark this answers is not in the document]
Comment (KennethEnevoldsen) [reply src/dfm/projects/data_processing/convert_danews_to_jsonlgz.py:32]: Det er forskelligt fra nyhedsmedie. Ofte er det bare Lead eller et udsnit fra bodytekst. Det kan skabe problemer for text redundancy. Fandt ud af det i sidste rundt at cleaning fordi teksterne var meget repetitative.

File: src/dfm/projects/data_processing/convert_danews_to_jsonlgz.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    example["metadata"] = metadata
+    return example
+
+
+def reformat_dataset(ds: Dataset, num_proc: int = 8) -> Dataset:
+    ds = ds.map(create_text, num_proc=num_proc)  # type: ignore
+    ds = ds.map(lambda x: {"source": "infomedia"}, num_proc=num_proc)  # type: ignore # noqa: ARG005
+    ds = ds.map(lambda x: {"added": "2022-10-24"}, num_proc=num_proc)  # type: ignore # noqa: ARG005
+    ds = ds.map(create_metadata, num_proc=num_proc)  # type: ignore
+    ds = ds.map(lambda x: {"id": x["ArticleId"]}, num_proc=num_proc)  # type: ignore
```
Comment (KennethEnevoldsen) [inline src/dfm/projects/data_processing/convert_danews_to_jsonlgz.py:60]:
```suggestion
```

File: src/dfm/projects/data_processing/convert_danews_to_jsonlgz.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    # current keys are:
+    # dict_keys(['ArticleUrl', 'Heading', 'SubHeading', 'Lead', 'Paragraph', 'PublishDate', 'BodyText', 'Captions', 'Authors', 'Source', 'WordCount', 'ArticleId', 'PageIds', 'Section'])  # noqa
+
+    # ids in file are not are not unique -->
+    # create new id:
+    ids = list(range(len(ds)))
+    ds = ds.add_column("id", ids)  # type: ignore
+
+    # reformat dataset
+    ds = reformat_dataset(ds)  # type: ignore
```
[the remark this answers is not in the document]
Comment (KennethEnevoldsen) [reply src/dfm/projects/data_processing/convert_danews_to_jsonlgz.py:100]: Jo det har du helt ret i. Fikser jeg lige.

File: src/dfm/projects/data_processing/twitter/conversations_to_threads.py [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+    }
+
+
+def created_at_to_timestamp(created_at: str) -> str:
+    """
+    Convert created_at to timestamp
+    created_at: "2021-04-13T12:52:46.000Z" -->
+    timestamp: "2021-04-13 12:52:46"
+    """
+    return created_at.replace("T", " ").replace(".000Z", "")
```
[the remark this answers is not in the document]
Comment (KennethEnevoldsen) [reply src/dfm/projects/data_processing/twitter/conversations_to_threads.py:191]: skulle vi i stedet bare køre med: 2021-04-13T12:52:46.000Z?

**Comment by KennethEnevoldsen** [comment]
Create a validator for the dataset format

**Review by peterbjorgensen — APPROVED** [review]
Looks good to me.

Notice that the parallel processing of Dolma works on file level, so the number of files should ideally be large enough to occupy a number of cores, i.e. it is not ideal to make a single big blob of data. I don't think it will be a problem, because we should have enough datasets to keep the cores busy, just something to keep in mind. It makes sense that a single file is the smallest unit for parallel processing because the gzip format only allows sequential access.

**Pull request merged**